### PR TITLE
Feature/add norwegian insurance certificates

### PR DESCRIPTION
**Pull request by AlexSJ1990**

### CODE REVIEW

File: src/main/java/com/hedvig/contract/service/InsuranceCertificateGeneratorV1.kt [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    val pdfData = pdfResponse.body
+    if (!pdfResponse.statusCode.is2xxSuccessful
+      || pdfResponse.headers.contentType != MediaType.APPLICATION_PDF
+      || pdfData == null || pdfData.size == 0) {
+//      log.error(
+//        "Failed to create PDF for product {}, content-types={} status={}",
+//        insuranceCertificate.agreement.id,
+//        pdfResponse.headers.contentType,
+//        pdfResponse.statusCode
+//      )
```
Comment (FredrikAreschoug) [inline src/main/java/com/hedvig/contract/service/InsuranceCertificateGeneratorV1.kt:71]: Remove?

File: src/main/java/com/hedvig/contract/service/InsuranceCertificateGeneratorV1.kt [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      throw RuntimeException("Html to PDF call to external service failed")
```
Comment (FredrikAreschoug) [inline src/main/java/com/hedvig/contract/service/InsuranceCertificateGeneratorV1.kt:72]: Maybe use the log error message here to get some more information on which certificate creation failed

File: src/main/java/com/hedvig/contract/service/InsuranceCertificateGeneratorV1.kt [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      inceptionDate = insuranceCertificate.inceptionDate
+    )
+  }
+
+  fun generateCertificateHtml(insuranceCertificate: InsuranceCertificateTemplate): List<String> {
+    return insuranceCertificate.getInsuranceCertHtmlFiles().map { file ->
+      val ctx = VelocityContext()
+      ctx.put("certificate", insuranceCertificate)
+      ctx.put("tools", VelocityTools())
+      val html = StringWriter()
```
Comment (FredrikAreschoug) [inline src/main/java/com/hedvig/contract/service/InsuranceCertificateGeneratorV1.kt:54]: I know nothing about `VelocityContext` and `VelocityTools` but is it something we want to recreate for each certificate?

File: src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateManager.kt [hunk 4]
```diff
@@ -1,64 +1,64 @@
...
+    holderMemberId: String,
+    isDraft: Boolean
+  ): InsuranceCertificateTemplate {
+    val member = memberService.getMember(holderMemberId)
+    return when(agreement) {
+      is SwedishApartmentAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
+      is SwedishHouseAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
+      is NorwegianHomeContentAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
+      is NorwegianTravelAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
+      else -> throw RuntimeException("")
```
Comment (FredrikAreschoug) [inline src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateManager.kt:61]: Some sort of error message would be nice :smile:

File: src/test/java/com/hedvig/contract/testUtils/BaseAgreementImpl.kt [hunk 5]
```diff
@@ -31,6 +32,15 @@ data class BaseAgreementImpl(
 ) {
   override fun getPreferredCurrency(): CurrencyUnit = Monetary.getCurrency("SEK")
 
+  override fun getInsuranceCertificate(
+    agreement: BaseAgreement,
+    inceptionDate: LocalDate,
+    member: InternalMember,
+    isDraft: Boolean
+  ): InsuranceCertificateTemplate {
+    throw Exception("test")
```
Comment (FredrikAreschoug) [inline src/test/java/com/hedvig/contract/testUtils/BaseAgreementImpl.kt:41]: test?

File: src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateTemplate.kt [hunk 6]
```diff
@@ -1,10 +1,42 @@
...
+  abstract val inceptionDate: LocalDate
+  abstract val holderMemberId: String
+  abstract val agreement: BaseAgreement
+  abstract val memberInfo: MemberInfo
+  abstract val isDraft: Boolean
+
+  fun getInsuranceCertHtmlFiles():List<String> {
+    return when(agreement) {
+      is NorwegianTravelAgreement -> listOf("test_cert1", "test_cert2")
+      is NorwegianHomeContentAgreement -> listOf("test_cert1", "test_cert2")
```
Comment (FredrikAreschoug) [inline src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateTemplate.kt:26]: should these be `test...`

File: src/test/java/com/hedvig/contract/testUtils/BaseAgreementImpl.kt [hunk 7]
```diff
@@ -38,7 +38,7 @@ data class BaseAgreementImpl(
     member: InternalMember,
     isDraft: Boolean
   ): InsuranceCertificateTemplate {
-    throw Exception("test")
+    TODO("not implemented")
```
Comment (meletios) [inline src/test/java/com/hedvig/contract/testUtils/BaseAgreementImpl.kt:41]: `IllegalStateException` fits better here or?

File: src/test/java/com/hedvig/productPricing/service/service/certificates/InsuranceCertificateGeneratorTest.kt [hunk 8]
```diff
@@ -212,9 +215,10 @@ class InsuranceCertificateGeneratorTest {
     assertThat(htmlCaptor.value).doesNotContain("Student")
   }
 
+
   @Test
   @Throws(IOException::class)
-  fun whenHouseAgreementReturnsHouseData() {
+  fun whenThereAreZeroCoinsuredsShouldNotShowCoInsuredText() {
```
Comment (meletios) [inline src/test/java/com/hedvig/productPricing/service/service/certificates/InsuranceCertificateGeneratorTest.kt:221]: Why we replace this test? Can we have both?

Comment (AlexSJ1990) [reply src/test/java/com/hedvig/productPricing/service/service/certificates/InsuranceCertificateGeneratorTest.kt:221]: whoops - didn't mean to replace it - will fix!

File: src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateTemplate.kt [hunk 9]
```diff
@@ -25,4 +28,29 @@ abstract class InsuranceCertificateTemplate {
...
+
+  fun getRenewalDate(): LocalDate {
+    return inceptionDate.plusYears(1).minusDays(1)
+  }
+
+  fun getLocalDate(): LocalDate {
+    return LocalDate.now()
+  }
+
+  fun getBasePremiumMultipliedBy12(): MonetaryAmount {
```
Comment (palmenhq) [inline src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateTemplate.kt:40]: i think we should name this what it is - `getYearlyBasePremium` or something?

File: src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateManager.kt [hunk 10]
```diff
@@ -41,7 +46,10 @@ class InsuranceCertificateManager @Autowired constructor(
     holderMemberId: String
   ): InsuranceCertificate {
     val insuranceCertificateDraft = when(agreement) {
-      is SwedishApartmentAgreement, is SwedishHouseAgreement -> {
+      is SwedishApartmentAgreement,
+      is SwedishHouseAgreement,
+      is NorwegianHomeContentAgreement,
+      is NorwegianTravelAgreement -> {
```
Comment (palmenhq) [inline src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateManager.kt:52]: Do we even have any other type of agreements? I think `getInsuranceCert` should be sufficient

File: src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateManager.kt [hunk 11]
```diff
@@ -64,8 +72,9 @@ class InsuranceCertificateManager @Autowired constructor(
     return when(agreement) {
       is SwedishApartmentAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
       is SwedishHouseAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
-      else -> throw RuntimeException("Expected agreement to be of type SwedishApartmentAgreement or SwedishHouseAgreement " +
-        "but was ${agreement.javaClass.simpleName}")
+      is NorwegianTravelAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
+      is NorwegianHomeContentAgreement -> agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
+      else -> throw RuntimeException("Unknown BaseAgreement: ${agreement::class.simpleName}")
```
Comment (palmenhq) [inline src/main/java/com/hedvig/contract/shared/certificates/InsuranceCertificateManager.kt:77]: And same here, i think
```kt
    return agreement.getInsuranceCertificate(agreement, inceptionDate, member, isDraft)
```

should be sufficient

File: src/main/resources/certificate/insurance_certificate_nor_travel.vm [hunk 12]
```diff
@@ -0,0 +1,1554 @@
...
+    <div class="grid__column grid__column--4/12">
+      <p>Eksklusiv forsikringsgiver for Hedvigs forsikring er
+        HDI Global Specialty SE, Sverige Filial, org. no. 516402-6345.
+        Hedvig er under tilsyn av det svenske finanstilsynet.
+      </p>
+    </div>
+  </div>
+</div>
+
+</div>
```
Comment (palmenhq) [inline src/main/resources/certificate/insurance_certificate_nor_travel.vm:1552]: Not that it'll affect the layout but we have a spare `</div>` here i think :)

File: src/main/java/com/hedvig/contract/type/norwegianTravel/certificates/NorwegianTravelCertificate.kt [hunk 13]
```diff
@@ -13,7 +13,7 @@ class NorwegianTravelCertificate(
   override val isDraft: Boolean
 ) : InsuranceCertificateTemplate() {
   override fun getInsuranceCertHtmlFiles(): String {
```
Comment (palmenhq) [inline src/main/java/com/hedvig/contract/type/norwegianTravel/certificates/NorwegianTravelCertificate.kt:15]: i guess this is really one `file` now not `files`?

**Pull request merged**